pkg/util/net/dial.go

```diff
@@ -0,0 +1,103 @@
+package net
+
+import (
+	"crypto/tls"
+	"net"
+)
+
+type dialOptions struct {
+	proxyURL                 string
+	protocol                 string
+	laddr                    string
+	addr                     string
+	tlsConfig                *tls.Config
+	disableCustomTLSHeadByte bool
+}
+
+type DialOption interface {
+	apply(*dialOptions)
+}
+
+type EmptyDialOption struct{}
+
+func (EmptyDialOption) apply(*dialOptions) {}
+
+type funcDialOption struct {
+	f func(*dialOptions)
+}
+
+func (fdo *funcDialOption) apply(do *dialOptions) {
+	fdo.f(do)
+}
+
+func newFuncDialOption(f func(*dialOptions)) *funcDialOption {
+	return &funcDialOption{
+		f: f,
+	}
+}
+
+func DefaultDialOptions() dialOptions {
+	return dialOptions{
+		protocol: "tcp",
+	}
+}
+
+func WithProxy(proxyURL string) DialOption {
```

Comment:
`WithProxy` -> `WithProxyURL`，params for `WithProxy` seems better if it's a new Proxy struct.